pdns/recursordist/rec-tcpout.hh

```diff
@@ -0,0 +1,73 @@
+/*
+ * This file is part of PowerDNS or dnsdist.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#pragma once
+
+#include "iputils.hh"
+#include "tcpiohandler.hh"
+
+class TCPOutConnectionManager
+{
+public:
+  // Max idle time for a connection, 0 is no timeout
+  static struct timeval maxIdleTime;
+  // Per thread maximum of idle connections for a specific destination, 0 means no idle connections will be kept open
+  static size_t maxIdlePerAuth;
+  // Max total number of queries to handle per connection, 0 is no max
+  static size_t maxQueries;
+  // Per thread max # of idle connections, 0 means no idle connections will be kept open
+  static size_t maxIdlePerThread;
+
+  struct Connection
+  {
+    std::string toString() const
+    {
+      if (d_handler) {
+        return std::to_string(d_handler->getDescriptor()) + ' ' + std::to_string(d_handler.use_count());
+      }
+      return "";
+    }
+
+    std::shared_ptr<TCPIOHandler> d_handler;
+    timeval d_last_used{0, 0};
+    size_t d_numqueries{0};
+  };
+
+  void store(const ComboAddress& ip, Connection& connection);
+  Connection get(const ComboAddress& ip);
+  void cleanup();
+
+  size_t size() const
+  {
+    return d_idle_connections.size();
+  }
+  uint64_t* getSize() const
+  {
+    return new uint64_t(size());
+  }
+
+private:
+  std::multimap<ComboAddress, Connection> d_idle_connections;
```

Comment:
One nit but this will probably not work well if we later want to use SNI and have different connections to the same host using different names. That's fine for now, though.

Reply:
Yep, I'll make a not of that in a comment.